relay-server/src/actors/envelopes.rs

```diff
@@ -765,6 +768,69 @@ impl EnvelopeProcessor {
         });
     }
 
+    /// Validates and extracts client reports.
+    ///
+    /// At the moment client reports are primarily used to transfer outcomes from
+    /// client SDKs.  The outcomes are removed here and sent directly to the outcomes
+    /// system.
+    fn process_client_reports(&self, state: &mut ProcessEnvelopeState) {
+        let mut timestamp = None;
+        let mut discarded_events = BTreeMap::new();
+        let received = state.envelope_context.received_at;
+
+        let clock_drift_processor = ClockDriftProcessor::new(state.envelope.sent_at(), received)
+            .at_least(MINIMUM_CLOCK_DRIFT);
+
+        // we're going through all client reports but we're effectively just merging
+        // them into the first one.
+        state.envelope.retain_items(|item| {
+            if item.ty() != ItemType::ClientReport {
+                return true;
+            };
+            match ClientReport::parse(&item.payload()) {
+                Ok(report) => {
+                    for (reason, category, quantity) in report.discarded_events.into_iter() {
+                        if reason.len() > 200 {
+                            relay_log::trace!("ignored client outcome with an overlong reason");
+                            continue;
+                        }
+                        *discarded_events.entry((reason, category)).or_insert(0) += quantity;
+                    }
+                    if let Some(ts) = report.timestamp {
+                        timestamp.get_or_insert(ts);
+                    }
+                }
+                Err(err) => relay_log::trace!("invalid client report received: {}", LogError(&err)),
+            }
+            false
+        });
+
+        let timestamp =
+            timestamp.get_or_insert_with(|| UnixTimestamp::from_secs(received.timestamp() as u64));
+
+        if clock_drift_processor.is_drifted() {
+            relay_log::trace!("applying clock drift correction to client report");
+            clock_drift_processor.process_timestamp(timestamp);
+        }
+
+        if discarded_events.is_empty() {
+            return;
+        }
+
+        let producer = OutcomeProducer::from_registry();
+        for ((reason, category), quantity) in discarded_events.into_iter() {
+            producer.do_send(TrackOutcome {
```

Comment:
This is effectively doing what the other outcome code already does, but i think this is not safe here now. The reason being that previously the number of outcomes was the same as the number of envelopes, so setting the outcome mailbox size to the envelope buffer size won't be sufficient any more:

https://github.com/getsentry/relay/blob/956e0b382a0f9917657d5cc7fcc7a613198e00a2/relay-server/src/actors/outcome.rs#L503-L504

Not entirely sure what out current approach for such things is. Just more buffer?

Reply:
We already send multiple outcomes per envelope when it comes to dropping event attachments and filtering, but there was a theoretical upper limit on the number of outcomes sent per envelope. This is unbounded. It would be good to define limits and start discarding reasons/categories if necessary to trim down the list

Reply:
Do we keep track of what happens if this overflows? Losing outcomes seems non great.

Reply:
After discussion this is fine because accepted is sent elsewhere (consumer) so we're good.